File: Semester 2/Algorithmen und Datenstrukturen/Praktikum/Blatt 01/Uebungsblatt1.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.Timer;
import java.util.stream.Collectors;

public class Uebungsblatt1 {
    private static class MaxTeilsumme {
        List<Integer> set;
        int firstIndex;
        int lastIndex;
        int sum;
        long additions;

        private MaxTeilsumme(List<Integer> set) {
            this.set = set;
            this.sum = Integer.MIN_VALUE;
            this.additions = 0;
        }

        public static MaxTeilsumme calculate(List<Integer> set) {
            final MaxTeilsumme result = new MaxTeilsumme(set);
            final int length = set.size();

            for (int i = 0; i < length; i++) {
                for (int j = i; j < length; j++) {
                    int sum = 0;
                    for (int k = i; k <= j; k++) {
                        sum += set.get(k);
                        result.additions++;
                    }
                    if (sum > result.sum) {
                        result.sum = sum;
                        result.firstIndex = i;
                        result.lastIndex = j;
                    }
                }
            }

            return result;
        }

        public List<Integer> getSet() {
            return set;
        }

        public int getFirstIndex() {
            return firstIndex;
        }

        public int getLastIndex() {
            return lastIndex;
        }

        public int getSum() {
            return sum;
        }

        public long getAdditions() {
            return additions;
        }
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        System.out.print("Dateiname: ");
        final String filename = sc.next();

        System.out.println(Path.of(filename).toAbsolutePath().toString());

        List<Integer> set;
        try (BufferedReader file = new BufferedReader(new FileReader(filename))) {
            set = file.lines().map((s) -> Integer.parseInt(s)).collect(Collectors.toList());
        }

        long start = System.nanoTime();
        MaxTeilsumme result = MaxTeilsumme.calculate(set);// ...
        long finish = System.nanoTime();
        double timeElapsed = (finish - start) / 1_000_000_000f;
        System.out.printf("Max. Teilsumme: %d\n", result.getSum());
        System.out.printf("Erster Index: %d\n", result.getFirstIndex());
        System.out.printf("Letzer Index: %d\n", result.getLastIndex());
        System.out.printf("Zeit in Sekunden: %f\n", timeElapsed);
        System.out.printf("Additionen: %d\n", result.getAdditions());
    }
}

File: Semester 2/Algorithmen und Datenstrukturen/Praktikum/Blatt 01/uebungsblatt1.py
import math
import time


def maxTeilsumme(a):
    n = len(a)
    maxsumme = -math.inf
    additionen = 0
    von = 0
    bis = 0
    for i in range(0, n):
        for j in range(i, n):
            summe = 0
            for k in range(i, j + 1):
                summe = summe + a[k]
                additionen = additionen + 1
            if summe > maxsumme:
                maxsumme = summe
                von = i
                bis = j
    return (von, bis, maxsumme, additionen)


a = list()
filename = input("Dateiname: ")
file = open(filename, mode='r')
for line in file:
    a.append(int(line))
file.close()

start = time.time()
(von, bis, maxsumme, additionen) = maxTeilsumme(a)
stop = time.time()

print("Max. Teilsumme", maxsumme)
print("Erster Index", von)
print("Letzter Index", bis)
print("Zeit in Sekunden", stop - start)
print("Additionen", additionen)
